src/envs/blocks.py

```diff
@@ -0,0 +1,439 @@
+"""Blocks domain. This environment IS downward refinable and DOESN'T
+require any backtracking. Its simplicity makes it a good testbed for
+predicate invention.
+"""
+
+from typing import List, Set, Sequence, Dict, Tuple, Optional
+import numpy as np
+from gym.spaces import Box
+from predicators.src.envs import BaseEnv
+from predicators.src.structs import Type, Predicate, State, Task, \
+    ParameterizedOption, Object, Action, GroundAtom, Image, Array
+from predicators.src.settings import CFG
+from predicators.src import utils
+
+
+class BlocksEnv(BaseEnv):
+    """Blocks domain.
+    """
+    # Parameters that aren't important enough to need to clog up settings.py
+    table_height = 0.2
+    x_lb = 1.3
+    x_ub = 1.4
+    y_lb = 0.15
+    y_ub = 20.85
+    held_tol = 0.5
+    clear_tol = 0.5
+    open_fingers = 0.8
+    pick_tol = 0.08
+    assert pick_tol < CFG.blocks_block_size
+    lift_amt = 1.0
+
+    def __init__(self) -> None:
+        super().__init__()
+        # Types
+        self._block_type = Type(
+            "block", ["pose_x", "pose_y", "pose_z", "held", "clear"])
+        self._robot_type = Type("robot", ["fingers"])
+        # Predicates
+        self._On = Predicate(
+            "On", [self._block_type, self._block_type], self._On_holds)
+        self._OnTable = Predicate(
+            "OnTable", [self._block_type], self._OnTable_holds)
+        self._GripperOpen = Predicate(
+            "GripperOpen", [self._robot_type], self._GripperOpen_holds)
+        self._Holding = Predicate(
+            "Holding", [self._block_type], self._Holding_holds)
+        self._Clear = Predicate(
+            "Clear", [self._block_type], self._Clear_holds)
+        # Options
+        self._Pick = ParameterizedOption(
+            # variables: [object to pick]
+            # params: [delta x, delta y, delta z]
+            "Pick", types=[self._block_type], params_space=Box(-1, 1, (3,)),
+            _policy=self._Pick_policy,
+            _initiable=self._Pick_initiable,
+            _terminal=self._Pick_terminal)
+        self._Stack = ParameterizedOption(
+            # variables: [object to stack currently held object onto]
```

Comment:
it took me like 30 seconds to parse this, can we say: "object on which to stack currently-held-object"